File: tests/test_Regularizor.cpp
/*
  alignlib - a library for aligning protein sequences

  $Id: test_Alignandum.cpp,v 1.3 2004/06/02 12:11:38 aheger Exp $

  Copyright (C) 2004 Andreas Heger

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

/** Test alignata objects
 */

#if HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <fstream>
#include <cassert>
#include <time.h>

#include "alignlib.h"
#include "alignlib_fwd.h"

#include "Alignandum.h"
#include "Encoder.h"
#include "HelpersEncoder.h"
#include "HelpersAlignandum.h"
#include "HelpersSubstitutionMatrix.h"

#define BOOST_TEST_MODULE
#include <boost/test/included/unit_test.hpp>
using boost::unit_test::test_suite;

using namespace std;
using namespace alignlib;

std::string ref_protein20 = "ACDEFGHIKLMNPQRSTVWY";
std::string ref_protein20x3 = ref_protein20 + ref_protein20 + ref_protein20;
std::string ref_protein20x3gaps = ref_protein20 + "-" + ref_protein20 + "Y" + ref_protein20 + "Y";

// most profilers only work with the twenty amino acid alphabet
// setDefaultEncoder( getEncoder(Protein20) );


void test_GenericRegularizor( const HRegularizor & r)
{
	getDefaultToolkit()->setRegularizor( r );
	HAlignandum a(makeProfile( ref_protein20x3, 3 ));
	a->prepare();
	std::cout << *a << std::endl;
}

void test_GenericRegularizorWithGaps( const HRegularizor & r)
{
	getDefaultToolkit()->setRegularizor( r );
	HAlignandum a(makeProfile( ref_protein20x3gaps, 3 ));
	a->prepare();
	std::cout << *a << std::endl;
}


/*
BOOST_AUTO_TEST_CASE( test_Regularizor )
{
	HRegularizor r( makeRegularizor());
	test_GenericRegularizor( r );
}
*/

BOOST_AUTO_TEST_CASE( test_RegularizorPsiblast )
{
  setDefaultEncoder( getEncoder( Protein20 ) );
  HRegularizor r( makeRegularizorPsiblast());
  test_GenericRegularizorWithGaps( r );
}

BOOST_AUTO_TEST_CASE( test_RegularizorDirichlet )
{
  setDefaultEncoder( getEncoder( Protein20 ) );
  HRegularizor r( makeRegularizorDirichletPrecomputed());
  test_GenericRegularizorWithGaps( r );
}

File: python/tests/test_Regularizor.py
# alignlib - a library for aligning protein sequences
# 
# $Id$
# 
# Copyright (C) 2004 Andreas Heger
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
import unittest, sys
from alignlib import *

class RegularizorCase( unittest.TestCase ):

    def setUp( self ):
        
        self.mFactory = makeRegularizor

        self.mMali = makeMultipleAlignment()
        
        self.mMali.add( makeAlignatum("ACDEFGHIKLMNPQRSTVWY") )
        self.mMali.add( makeAlignatum("ACDEFGHIKLMNPQRSTVWY") )
        self.mMali.add( makeAlignatum("ACDEFGHIKLMNPQRSTVWY") )
        
    def testRegularizor( self ):
                            
        # getDefaultToolkit().setEncoder( getEncoder( Protein20) )
        getDefaultToolkit().setRegularizor( self.mFactory() )
        a = makeProfile(self.mMali)
        a.prepare()
        print a

class RegularizorDirichletCase( RegularizorCase ):

    def setUp( self ):
        RegularizorCase.setUp( self )
        self.mFactory = makeRegularizorDirichletPrecomputed

"""
class WeightorHenikoffCase( WeightorCase ):

    def setUp( self ):
        self.mWeightorFactory = makeWeightorHenikoff

class WeightorHenikoffKimmenCase( WeightorCase ):

    def setUp( self ):
        self.mWeightorFactory = makeWeightorHenikoffKimmen
"""     

class TatusovCase( unittest.TestCase ):

    def testPsiblast( self ):

        matrix = makeSubstitutionMatrixBlosum62()

        bg = (0.078047, 0.053640, 0.062949, 0.038556, 0.038556,                                                                                                                                                              
                0.073772, 0.021992, 0.051420, 0.057438, 0.090191,                                                                                                                                  
                0.022425, 0.044873, 0.052028, 0.042644, 0.051295,                                                                                                                                  
                0.071198, 0.058413, 0.064409, 0.013298, 0.032165 )

        b = FrequencyVector()
        b.extend( bg )
        regularizor = makeRegularizorTatusov(
            matrix,
            b,
            "ACDEFGHIKLMNPQRSTVWY",                                                                                                                                                                                              
            10, 
            0.3176 )

    def testDNA( self ):

        matrix = makeSubstitutionMatrixDNA4()

        bg = (0.4, 0.3, 0.2, 0.2)

        weight_matrix = MatrixDouble( 10, 5, 0.0 )

        for x in range(0, 10):
            weight_matrix.setValue(x, x % 5, 1.0 )

        printMatrix( sys.stdout, weight_matrix )

        frequency_matrix = MatrixDouble( 10, 5, 0.0 )

        b = FrequencyVector()
        b.extend( bg )
        regularizor = makeRegularizorTatusov(
            matrix,
            b,
            "ACGT",                                                                                                                                                                                              
            10.0, 
            0.0 )

        regularizor.fillFrequencies( frequency_matrix,
                                     weight_matrix,
                                     getEncoder( DNA4) )
        
        printMatrix( sys.stdout, frequency_matrix )


def suite():
    suite = unittest.TestSuite()
    suite.addTest(RegularizorDirichletCase)
    suite.addTest(TatusovCase)
    return suite

if __name__ == "__main__":
    unittest.main()

